Allow overwrite archives extraction

isbm opened the pull request:
### What does this PR do?

Add an option `overwrite` to `states.archive`.

### Previous Behavior

If an archive was extracted, newer version was not applied.

### New Behavior

If an archive was extracted, all extracted content is flushed and an archive is extracted again.

### Tests written?

No

Commit:
Remove nonsense comment and react on generally absent path name

Commit:
Add overwrite option so the extraction of the archive can be always performed.

cachedout commented on salt/states/archive.py:
```diff
@@ -898,7 +903,14 @@ def extracted(name,
+    if extraction_needed:
+        destination = os.path.join(name, contents['top_level_dirs'][0])
+        if os.path.exists(destination):
+            shutil.rmtree(destination)
```
Try/except a failure around `rmtree` here?

cachedout commented:
I see where you're going here, but I'm a little concerned that this might have some unexpected side-effects. Say, for example, that an application is unzipped using this state on deployment but then it stores some stateful information inside the unzipped dir. Deleting the dir might nuke that which would be surprising.

Perhaps I'm being too conservative here but it did come to mind. Let me know what you think.

isbm commented:
@cachedout Sure it wil be nuked out. That's the feature is all about: to nuke what was before. :wink: But your use case is quite rare and if you _know_ you are doing this, then probably just never turn it to `True`. But since in a daily life it is not a case most of the time and _usually_ it is a good idea to unzip temporary archive somewhere in `/tmp` but the whole thing simply forbids that... then this leads to a pretty cumbersome usecases where you should track down your `/tmp/dynamicallycreateddirectories`.

cachedout commented:
@isbm Sure, I agree it's rare. It appears undocumented, though. What if we just documented it so that there were no surprises?

isbm commented:
@cachedout Seems reasonable. :+1:

Commit:
Prevent crash during externally changed archive permissions

Commit:
Document the behaviour.

isbm commented:
@cachedout Here you go:

- Documented the behaviour that it will flush the archive.
- Added exception handling.

cachedout commented:
@isbm Most excellent. Much appreciated!

isbm commented:
> You're assuming that these operations will be in the same SLS file. You don't know that they will be.
> We have a long-standing philosophy of trying to stay out of our users' way when we can. Expecting them to conform their SLS to a specific usage, just because you desire a certain behavior, goes against that.

@terminalmage I do not conform anything but managing my systems found the usage inconvenient and cumbersome, therefore offered a pre-flush method to ease pain working with this. As a user, you are totally free not to use it and back to where you were. But I as a user am not free to use it, because this was not implemented. I also fail to find an answer to my question, which was this as per **your** use-case:

1. Archive is going to be extracted to the `$DESTINATION`
2. A config file is going to be added to the `$DESTINATION` after the extraction of the archive (NOTE: `archive.extracted` does not have such feature, so another operation is needed _anyway_).
3. An additional formula may be ran over earlier mentioned SLS file. This step is optional, but I would expect that too.

With the `overwrite=True` the scenario above is perfectly repeatable. With the `overwrite=False` the scenario above is impossible, because to make it repeatable, one would need to add an additional check for the `$DESTINATION` be properly removed and also added to the `archive.extracted`, otherwise `archive.extracted` will fail. This requires more code and therefore more potential bugs. This is why usual workaround a good-old `cmd.run` comes for help in these cases.

> If we want to remove paths before extracting, this should be something that is configurable, not something that the user is forced into doing.

The parameter is set to `False` by default. This does not forces anything, in my eyes. If nobody changes that to `overwrite=True`, nothing special will happen. But if one needs the old place be overwritten, however `archive.extracted` says "No!", then _this_ is forced to no-op, as I see it.
